README.md
# SRI for broccoli
[![build status](https://secure.travis-ci.org/jonathanKingston/broccoli-sri-hash.svg)](http://travis-ci.org/jonathanKingston/broccoli-sri-hash)
[![build status](https://ci.appveyor.com/api/projects/status/github/jonathanKingston/broccoli-sri-hash?branch=master&svg=true)](https://ci.appveyor.com/project/jonathanKingston/broccoli-sri-hash/branch/master)
[![npm status](http://img.shields.io/npm/v/broccoli-sri-hash.svg)](https://www.npmjs.org/package/broccoli-sri-hash)
[![dependency status](https://david-dm.org/jonathanKingston/broccoli-sri-hash.svg)](https://david-dm.org/jonathanKingston/broccoli-sri-hash)

This plugin looks at an apps html files to rewrite their content with integrity attributes.

### Options

- **origin** - if `crossorigin` isn't specified but `prepend` is it will add an integrity if `prepend` starts with `origin`
- **crossorigin** - adds a crossorigin attribute to script and link elements
    - This is **required** for CORS resources values are:
        - `use-credentials`
        - `anonymous`
- **prepend** - resources with a full path will only get an applied integrity if the md5 checksum passes
- **paranoiaCheck** - false by default, this turns off the integrity attribute if any Unicode is found within the file.

### Example
```js
var sriTree = sri('path/to/code', {
  prefix: 'https://example.com/',
  crossorigin: 'anonymous'
});
```

### 'paranoiaCheck'

There was an encoding issue based on certain characters which is [landed in Chrome 46](https://code.google.com/p/chromium/issues/detail?id=527286) when using Chrome.
This check fails if there is any non ASCII characters. On failure the file won't have a integrity attribute added.
**Please note** this will be removed as a default in the future; with the desire to remove all of the checking code too.


CHANGELOG.md
# 2.1.1

- Updating linting and snyk dev dependency packages
  - eslint@2.5.1
  - eslint-config-nightmare-mode@2.3.0
  - mocha-eslint@2.0.2
  - snyk@1.13.1

# 2.1.0

- Adding in fixes for utf8 streams
- Adding fingerprintCheck for disabling md5 file checking

# 2.0.0

- Removing paranoiaCheck as a default

# 1.2.2

- Removal of snyk as came out in ember setups

# 1.2.1

- Base tag fixes

# 1.2.0

- Update to use broccoli-caching-writer as a base class. Fixes issues with more aggressive caching added in recent broccoli-filter versions.

# 1.1.1
- @kellyselden fixes for node 0.10
  -  windows cmd fix
  -  adding promise polyfill for node 0.10
  -  adding appveyor badge
  -  mirroring ember-cli travis envs

# 1.1.0
- Added in a temporary fix for Unicode encoding issue which disables the code if the library detects unicode.

# 1.0.4
- Make the filter dependency in the dependency setup

# 1.0.3
- Fixing extra closing slash and whitespace in output
- Update to latest broccoli-filter

# 1.0.2
- Changing packages to be semver compatible

# 1.0.1
- Changed code to work with 0.2.x broccoli-filter
- Made stateless checks be global to remove recalculation
- Adding in more failure tests for broken HTML
- Cleaning up regex to cope with failure cases

# 1.0.0
- Added in build steps and badges

# 0.2.4
- Made code only work with stylesheet link elements

# 0.2.3
- Upgrade to the latest toolbox

# 0.2.2
- Added origin attribute checking to simplify interface


test/index.js
var chai = require('chai');
var assert = chai.assert;
var fs = require('fs');
var broccoli = require('broccoli');
var plugin = require('../');
var lint = require('mocha-eslint');
var walkSync = require('walk-sync');

function file(path) {
  return fs.readFileSync(path,  'UTF-8').trim();;
}

describe('broccoli-sri-hash', function () {
  var builder;

  before(function() {
    builder = new broccoli.Builder(plugin('test/fixtures/input', {
      prefix: 'https://example.com/',
      crossorigin: 'anonymous'
    }));
  });

  after(function() {
    builder.cleanup();
  });

  it('rule outputs match (initial build)', function () {
    return builder.build().then(function(output) {
      var actual, expected;

      actual = file(output.directory + '/test.html');
      expected = file('test/fixtures/output/test.html');

      assert.equal(actual, expected);

      actual = file(output.directory + '/foo/test.html');
      expected = file('test/fixtures/output/foo/test.html');

      assert.equal(actual, expected);

      actual = file(output.directory + '/two/deep/test.html');
      expected = file('test/fixtures/output/two/deep/test.html');

      assert.equal(actual, expected);

      assert.deepEqual(walkSync(output.directory), [
        'favicon.png',
        'foo/',
        'foo/test.html',
        'moment-with-locales.min.js',
        'omg.png',
        'other.css',
        'test.html',
        'thing.js',
        'two/',
        'two/deep/',
        'two/deep/test.html',
        'unicode-chars.js'
      ]);
    });
  });

  it('rule outputs match (rebuild)', function () {
    var pathToMutate = 'test/fixtures/input/other.css';
    var originalContent = fs.readFileSync(pathToMutate);
    return builder.build().then(function(output) {
      // mutate input File
      fs.writeFileSync('test/fixtures/input/other.css', '* { display: none; }');

      assert.deepEqual(walkSync(output.directory), [
        'favicon.png',
        'foo/',
        'foo/test.html',
        'moment-with-locales.min.js',
        'omg.png',
        'other.css',
        'test.html',
        'thing.js',
        'two/',
        'two/deep/',
        'two/deep/test.html',
        'unicode-chars.js'
      ]);
      return builder.build();
    }).then(function(output) {
      var actual = file(output.directory + '/test.html');
      var expected = file('test/fixtures/output2/test.html');

      assert.deepEqual(walkSync(output.directory), [
        'favicon.png',
        'foo/',
        'foo/test.html',
        'moment-with-locales.min.js',
        'omg.png',
        'other.css',
        'test.html',
        'thing.js',
        'two/',
        'two/deep/',
        'two/deep/test.html',
        'unicode-chars.js'
      ]);
      assert.equal(actual, expected);
    }).finally(function() {
      fs.writeFileSync(pathToMutate, originalContent);
    });
  });

});


lint([
  'index.js',
  'tests/index.js'
]);


index.js
var CachingWriter = require('broccoli-caching-writer');
var sriToolbox = require('sri-toolbox');
var fs = require('fs');
var crypto = require('crypto');
var symlinkOrCopy = require('symlink-or-copy').sync;
var Promise = require('rsvp').Promise; // node 0.10
var path = require('path');

var STYLE_CHECK = /\srel=["\'][^"]*stylesheet[^"]*["\']/;
var SRC_CHECK = /\ssrc=["\']([^"\']+)["\']/;
var HREF_CHECK = /\shref=["\']([^"\']+)["\']/;
var SCRIPT_CHECK = new RegExp('<script[^>]*src=["\']([^"]*)["\'][^>]*>', 'g');
var LINT_CHECK = new RegExp('<link[^>]*href=["\']([^"]*)["\'][^>]*>', 'g');
var INTEGRITY_CHECK = new RegExp('integrity=["\']');
var CROSS_ORIGIN_CHECK = new RegExp('crossorigin=["\']([^"\']+)["\']');
var MD5_CHECK = /^(.*)[-]([a-z0-9]{32})([.].*)$/;
var mkdirp = require('mkdirp');

function SRIHashAssets(inputNodes, options) {
  if (!(this instanceof SRIHashAssets)) {
    return new SRIHashAssets(inputNodes, options);
  }

  this.options = options || {};
  this.context = this.options.context || {};
  var nodes = inputNodes;
  if (!Array.isArray(nodes)) {
    nodes = [nodes];
  }

  CachingWriter.call(this, nodes, {
    // disabled to ensure all files are synced forward
    // I suspect additions to BCW are needed, or a slightly different plugin
    // to handle this more elegantly.
    // Leaving this comment here as a reminder. -sp
    //
    // cacheInclude: [
    //   /\.html$/,
    //   /\.js$/,
    //   /\.css$/
    // ]
  });

  if (!('paranoiaCheck' in this.options)) {
    this.options.paranoiaCheck = false;
  }

  if (!('fingerprintCheck' in this.options)) {
    this.options.fingerprintCheck = true;
  }

  if ('origin' in this.options) {
    if ('prefix' in this.options && !('crossorigin' in this.options)) {
      if (this.options.prefix.indexOf(this.options.origin, 0) === 0) {
        this.options.crossorigin = false;
      }
    }
  }
}

SRIHashAssets.prototype = Object.create(CachingWriter.prototype);
SRIHashAssets.prototype.constructor = SRIHashAssets;

SRIHashAssets.prototype.addSRI = function addSRI(string) {
  var plugin = this;

  return string.replace(SCRIPT_CHECK, function srcMatch(match) {

    var src = match.match(SRC_CHECK);
    var filePath;

    if (!src) {
      return match;
    }

    filePath = src[1];

    return plugin.mungeOutput(match, filePath);
  }).replace(LINT_CHECK, function hrefMatch(match) {
    var href = match.match(HREF_CHECK);
    var isStyle = STYLE_CHECK.test(match);
    var filePath;

    if (!isStyle || !href) {
      return match;
    }

    filePath = href[1];

    return plugin.mungeOutput(match, filePath);
  });
};

SRIHashAssets.prototype.readFile = function readFile(file) {
  var assetSource;

  try {
    assetSource = fs.readFileSync(this.inputPaths[0] + '/' + file, 'utf8');
  } catch(e) {
    return null;
  }

  return assetSource;
};

/*
  If 'paranoiaCheck' is enabled then it will check a file only contains ASCII characters
    - will return true if paranoiaCheck is disabled
    - will return true if ASCII only
    - will return false if non ASCII chars are present
  This relates to an issue that is either within OpenSSL or Chrome itself due to an encoding issue:
   https://code.google.com/p/chromium/issues/detail?id=527286
*/
SRIHashAssets.prototype.paranoiaCheck = function paranoiaCheck(assetSource) {
  var i;
  var checkResult = true;

  if (this.options.paranoiaCheck === true) {
    for (i = 0; i < assetSource.length; i++) {
      if (assetSource.charCodeAt(i) > 127) {
        checkResult = false;
        break;
      }
    }
  }

  return checkResult;
};

SRIHashAssets.prototype.generateIntegrity = function generateIntegrity(output, file, external) {
  var assetSource = this.readFile(file);
  var selfCloseCheck = /\s*\/>$/;
  var integrity;
  var append;
  var outputWithIntegrity;

  if (assetSource === null) {
    return output;
  }

  if (this.paranoiaCheck(assetSource) === false) {
    return output;
  }


  integrity = sriToolbox.generate({
    algorithms: ['sha256', 'sha512']
  }, assetSource);

  append = ' integrity="' + integrity + '"';

  if (external && this.options.crossorigin) {
    if (!CROSS_ORIGIN_CHECK.test(output)) {
      append = append + ' crossorigin="' + this.options.crossorigin + '" ';
    }
  }

  if (selfCloseCheck.test(output)) {
    outputWithIntegrity = output.replace(selfCloseCheck, append + ' />');
  } else {
    outputWithIntegrity = output.replace(/\s*[>]$/, append + ' >');
  }
  return outputWithIntegrity;
};

SRIHashAssets.prototype.checkExternal = function checkExternal(output, file) {
  var md5Matches = file.match(MD5_CHECK);
  var md5sum = crypto.createHash('md5');
  var assetSource;
  var filePath;

  if (!('prefix' in this.options) || !('crossorigin' in this.options) || md5Matches === null) {
    return output;
  }

  filePath = file.replace(this.options.prefix, '');

  if (filePath === file) {
    return output;
  }

  assetSource = this.readFile(filePath);
  if (assetSource === null) {
    filePath = md5Matches[1].replace(this.options.prefix, '') + md5Matches[3];
    assetSource = this.readFile(filePath);
    if (assetSource === null) {
      return output;
    }
  }

  md5sum.update(assetSource, 'utf8');
  if (this.options.fingerprintCheck === false || md5Matches[2] === md5sum.digest('hex')) {
    return this.generateIntegrity(output, filePath, true);
  }
  return output;
};

SRIHashAssets.prototype.mungeOutput = function mungeOutput(output, filePath) {
  var newOutput = output;

  if (/^https?:\/\//.test(filePath)) {
    return this.checkExternal(output, filePath);
  }

  if (!INTEGRITY_CHECK.test(output)) {
    newOutput = this.generateIntegrity(output, filePath);
  }

  return newOutput;
};

SRIHashAssets.prototype.processHTMLFile = function processFile(entry) {
  var fileContent = this.addSRI(fs.readFileSync(entry.fullPath, 'utf8'));
  var fullPath = this.outputPath + '/' + entry.relativePath;

  mkdirp.sync(path.dirname(fullPath));

  fs.writeFileSync(fullPath, fileContent);
};

SRIHashAssets.prototype.processOtherFile = function(entry) {
  var fullPath = this.outputPath + '/' + entry.relativePath;
  mkdirp.sync(path.dirname(fullPath));
  symlinkOrCopy(entry.fullPath, fullPath);
};

SRIHashAssets.prototype.build = function () {
  var html = [];
  var other = [];

  this.listEntries().forEach(function(entry) {
    if (/\.html$/.test(entry.relativePath)) {
      html.push(entry);
    } else {
      other.push(entry);
    }
  });

  return Promise.all([
    Promise.all(html.map(this.processHTMLFile.bind(this))),
    Promise.all(other.map(this.processOtherFile.bind(this)))
  ]);
};

module.exports = SRIHashAssets;
